Test Improvements & Benchmarks for SqlKata Compiler

faddiv opened the pull request:
Hi Ahmad,

We've been using SqlKata in our project, and I’ve started exploring its internals. 
I thought the compiler part could be improved significantly both for speed and memory usage.
However, to rewrite it safely, I first wanted to improve the tests and add some benchmarks.

This PR shows what I did in this regard. Please review this PR, and let me know if you are interested in any part of my work. (I also have a branch in which I implemented a new compiler, just enough to run the benchmarks.)

### How to run the benchmark

In IDE:
- Set QueryBuilder.Benchmarks as the startup project.
- Set Build mode to release. (BanchmarkDotNet only runs on release compilation)
- Start the project without debugging.
Command line:
From the solution folder (on Windows):
`dotnet run -c Release --project .\QueryBuilder.Benchmarks\QueryBuilder.Benchmarks.csproj`

The reports are generated into the `.\BenchmarkDotNet.Artifacts\` folder at the solution level.

Meanwhile, it is recommended to close everything else, and do as little as possible so other processes don't affect the results.

### Summary of Changes:
- I broke down the tests that check the same input with multiple engines, with `Theory` and `InlineData` attributes. This way it can be seen if anything fails separately, and better visible if any test is missing.
   - At some places, I didn't do this breakdown everywhere since I was unsure if it was really worth it. (Base class functionality was tested.)
   - I replaced `TestCompilersContainer` completely. I was not sure in which direction you wanted to move these tests, but I saw somewhere that you wished to deprecate them.
- Created separate tests for the bindings.
- Created some benchmarks with three example queries.
   - These benchmarks only test the compiler part.
   - In benchmarks, I usually create pre-tests to validate whether they return the correct results, which I did here as well.
- Also, it was my goal, to make the compiler easily replaceable in the tests.

Looking forward to your feedback!

ahmad-moussawi committed:
add ci workflow

ahmad-moussawi committed:
remove Program folder

ahmad-moussawi committed:
chore: add release pipeline

ahmad-moussawi committed:
chore: update release ci to add nuget source

ahmad-moussawi committed:
chore: update readme

ahmad-moussawi commented:
Hey @faddiv thanks for your contribution, I would take a bit of time to review it, but it seems interesting, 
can you please add to the PR description how can I run the benchmarks?

faddiv commented:
Hi @ahmad-moussawi ,

I added the description. It simply can run with the following command: `dotnet run -c Release --project .\QueryBuilder.Benchmarks\QueryBuilder.Benchmarks.csproj`

ahmad-moussawi committed:
Add GitHub Actions workflow for .NET build and test

This workflow automates the build and test process for a .NET project on push and pull request events to the main branch.

Commit:
run ci on PRs

Commit:
reorganize ci.yml

ahmad-moussawi commented:
@faddiv sorry for the late reply, 
I have few comments on the tests adjustments.
Is it easy for you to address the Benchmarks in a separate PR, so we can merge it, then we discuss the tests enhancements?

ahmad-moussawi committed:
Add MS SQL configuration to CI workflow

ahmad-moussawi committed:
Change MSSQL user from 'admin' to 'sa'

ahmad-moussawi committed:
Replace MS SQL setup with installation commands

ahmad-moussawi committed:
Configure SQL Server service in CI workflow

Added SQL Server service container and updated installation steps for sqlcmd tools.

ahmad-moussawi committed:
Update ci.yml

ahmad-moussawi committed:
Update ci.yml

Commit:
use secret password for db

faddiv commented:
> @faddiv sorry for the late reply, I have few comments on the tests adjustments. Is it easy for you to address the Benchmarks in a separate PR, so we can merge it, then we discuss the tests enhancements?

Hi @ahmad-moussawi.

I will close this and reorganize this into two separate pr.
